cassandane/Cassandane/Cyrus/SearchFuzzy.pm

```diff
@@ -1926,5 +1920,305 @@ sub test_reindex_mb_uniqueid
+        mime_boundary => "123456789abcdef",
+        body => ""
+        ."\r\n--123456789abcdef\r\n"
+        ."Content-Type: text/plain\r\n"
+        ."\r\n"
+        ."bodyterm"
+        ."\r\n--123456789abcdef\r\n"
+        ."Content-Type: application/pdf\r\n"
+        ."\r\n"
+        ."attachterm"
+        ."\r\n--123456789abcdef--\r\n"
+    ) || die;
+
+    xlog "Run squatter with cachedir $cachedir";
+    $self->{instance}->run_command({cyrus => 1},
+        'squatter', "--attachextract-cache-dir=$cachedir", @squatterArgs);
+}
+
+sub test_squatter_attachextract_cache
+    :min_version_3_9 :needs_search_xapian :SearchAttachmentExtractor
+{
+    my ($self) = @_;
+    my $instance = $self->{instance};
+    my $imap = $self->{store}->get_client();
+
+    my $tracedir = tempdir (DIR => $instance->{basedir} . "/tmp");
+    $self->start_echo_extractor(tracedir => $tracedir);
+
+    xlog "Create and index index messages";
+    my $cachedir = tempdir (DIR => $instance->{basedir} . "/tmp");
+    $self->squatter_attachextract_cache_run($cachedir);
+
+    xlog "Assert text bodies of both messages are indexed";
+    my $uids = $imap->search('fuzzy', 'body', 'bodyterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert attachements of both messages are indexed";
+    $uids = $imap->search('fuzzy', 'xattachmentbody', 'attachterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert extractor only got called once";
+    my @tracefiles = glob($tracedir."/*_PUT_*");
+    $self->assert_num_equals(1, scalar @tracefiles);
+
+    xlog "Assert cache contains one file";
+    my @files = glob($cachedir."/*");
+    $self->assert_num_equals(1, scalar @files);
+}
+
+sub test_squatter_attachextract_cachedir_noperm
+    :min_version_3_9 :needs_search_xapian :SearchAttachmentExtractor :NoCheckSyslog
+{
+    my ($self) = @_;
+    my $instance = $self->{instance};
+    my $imap = $self->{store}->get_client();
+
+    my $tracedir = tempdir (DIR => $instance->{basedir} . "/tmp");
+    $self->start_echo_extractor(tracedir => $tracedir);
+
+    xlog "Run squatter with read-only cache directory";
+    my $cachedir = tempdir (DIR => $instance->{basedir} . "/tmp");
+    chmod "400", $cachedir;
+    $self->squatter_attachextract_cache_run($cachedir, "--allow-partials");
+
+    xlog "Assert text bodies of both messages are indexed";
+    my $uids = $imap->search('fuzzy', 'body', 'bodyterm');
+    $self->assert_deep_equals([1,2], $uids);
+
+    xlog "Assert attachements of both messages are not indexed";
+    $uids = $imap->search('fuzzy', 'xattachmentbody', 'attachterm');
+    $self->assert_deep_equals([], $uids);
+
+    xlog "Assert extractor got called twice with attachment uploads";
+    my @tracefiles = glob($tracedir."/*_PUT_*");
+    $self->assert_num_equals(2, scalar @tracefiles);
+
+    xlog "Assert cache contains no file";
+    chmod "700", $cachedir;
+    my @files = glob($cachedir."/*");
+    $self->assert_num_equals(0, scalar @files);
```

Comment:
One of these tests is leaving behind a directory with junk permissions that cassandane's --cleanup feature can't clean up, so the next cassandane run spews a bunch of permissions errors at startup.  I'm not sure if it's this test in particular, but it's is the one I see fiddling with permissions.

Here's the contents of the instance directory later, when everything has been cleaned up except the directory with the bad perms.  Note that owner has write, but no read or execute, group has write, and other has "T" (??):

```
$find /dev/shm/cass/012020011AD/ -ls
 19443457      0 drwxr-xr-x   3 cyrus    cyrus          60 Apr 19 11:41 /dev/shm/cass/012020011AD/
 19443480      0 drwxr-xr-x   3 cyrus    cyrus          60 Apr 19 11:41 /dev/shm/cass/012020011AD/tmp
 19443619      0 d-w-rwxr-T   2 cyrus    cyrus          40 Apr 19 11:41 /dev/shm/cass/012020011AD/tmp/PAF84XZC4r
```

I suspect the problem might be `chmod "400"` and `chmod "700"`, which I think maybe should be `chmod "0400"` and chmod `"0700"`?